### ISSUE TITLE
MapKnitterExporter architecture discussion

### ISSUE THREAD

**Issue by jywarren** [publiclab/mapknitter#298]
We are exploring parallel tracks for cloud-based MapKnitter exporting, and one option is a JavaScript based process. 

The base idea is to run the export process as a scalable web service, possibly "serverless" or REST, in Google Cloud and/or other cloud providers like Amazon AWS Lambda (primarily Google Cloud but compatible with others). Comments/suggestions/eurekas welcome! 🎉 

Importantly, either track would ideally present the same API so that we could compare their performance. 

## JavaScript track

In this track, more experimentally, we'd use [Image Sequencer](https://github.com/publiclab/image-sequencer/), possibly with the [webgl-distort](https://github.com/jywarren/webgl-distort) library.

The major challenges here, I'd guess, would be:

1. handling very big image files (up to 8mb each?) in memory
2. serious speed improvements in IS, such as the proposed WebAssembly or WebGL adapters
3. figuring out the best way to persist images for later access, and how to integrate the exporter with this (passing a callback function to upload them to a given store? Credentials?)
4. trying to duplicate or integrate GDAL's generation of a giant combined GeoTIFF (just really huge images to manage in memory?)
5. trying to duplicate or integrate GDAL's generation of TMS-formatted map tiles

For these last two, see #296 where there are some JS options to experiment with. 

Also, we would try to develop this track in such a way as to make it possible to run locally in the browser, natively or in an Electron-style local JS app.

## Ruby/ImageMagick/GDAL track

A more traditional approach is being explored here: #258, where we take the exporting sections currently featured in MapKnitter, and duplicate them in a minimal Ruby container that can be run on-demand.

## Spec

To guide the development of both tracks, we're imagining a basic common behavior of:

1. receiving a collection of image URLs or data-URLs of images AND a scale (cm/px or final pixel size)
2. outputting a combined JPG image at a given scale or pixel size
3. advanced versions might cut tiles or output GeoTiffs (see challenges in JS version above)

Links and resources are being compiled here: https://github.com/publiclab/mapknitter/issues/296

****

What have I missed? @tech4GT @icarito would you mind adding any questions, clarifications?

**Comment by tech4GT** [comment]
@jywarren this looks really nice! Things immediately make a lot more sense!🎉

**Comment by icarito** [comment]
I've just spent some time deploying a learning project to Google Cloud Platform (App Engine) as a Docker container. I've got a better understanding now of what is required! Thanks!

**Comment by jywarren** [comment]
OK, i'd like to add in an overview of the export system step by step; i've left notes where we might make changes or improvements as well, and will link to lines of code where these things currently happen!

Also -- a couple ideas:

1. Idea: produce separate GeoTiffs to skip the 
2. Idea: produce TMS tiles from any collection of images, given tile coordinates and image sources (with known corner coordinates)

****

# Breaking down the export process

## Separable steps

1. collect set of image URLs and their corner coordinates
2. for each image (could do this from existing Ruby code or in npm module):
    * determine image pixel dimensions
    * convert corner coordinates to pixel positions
3. for each image (using existing Ruby/ImageMagick code or in remote Image Sequencer container):
    * perform image distortion from original dimensions to new pixel corner positions
    * embed exif data for corner coordinates
    * save and return URL for download
    * (optional, could do later) produce GeoTiffs of each image
    * (optional, could do later) produce TMS tileset of each image
4. given collection of warped images, calculate pixel positions of image collection relative to each other (Ruby code exists)
    * (optional alternative) produce SVG or PDF containing images at relative positions (less memory use)
    * produce composite/merged image using this data
    * save and return URL of combined image for download
    * (optional) produce GeoTiff of combined image
    * (optional) pass GeoTiff to GDAL for conversion into traditional TMS tileset
5. Possible next steps:
    * produce merged TMS of step 3 per-image TMS tiles instead of generating from step 4's giant GeoTiff
    * produce single TMS from combination of per-image GeoTiffs from end of step 3

**Comment by jywarren** [comment]
@SidharthBansal @tech4GT @icarito just so you see this additional note breaking down the export process. There are portions that could be accomplished with traditional ImageMagick/GDAL combo just breaking out the Ruby-controlled code in our codebase (see #296 but i'll copy in more here), but I am hoping we can accomplish a lot in stand-alone containers in a serverless or at least remote REST model.

**Comment by tech4GT** [comment]
Starting work now! @icarito Can you please share some of the resources you have been going through, that would be a big help for me :)

**Comment by tech4GT** [comment]
@jywarren @icarito I would be starting with a basic express configuration that takes an image url and a sequencer string and returns the final output, can we create a repository for this on publiclab? Or should I make this on my github??

**Comment by tech4GT** [comment]
Okay a couple of things here
- We should add a flag to the run config which allows us to disable the progress logs(it'll unnecessarily slow down the server otherwise)
- I have some ideas in mind to speed up the pixelManipulation API which will in turn speed up most modules
- Should we return the output as a data uri or us the imgur service like we originally planned? Or maybe we can have a parameter in the request which allows both options
```js
/* Request Body */
{
'url': String, // URL if input image
'sequence': String // The sequence string which will be imported into sequencer,
'upload': boolean // Denotes whether to return the data uri or to upload to imgur and return that
}
```
How does this sound @jywarren @icarito ??

**Comment by jywarren** [comment]
How about https://github.com/publiclab/image-sequencer-app, which we
created a while back?

Let's start with a dataurl, but we should also plan to have an abstract way
to "put" the image somewhere.

Great!

**Comment by tech4GT** [comment]
@jywarren Ok pushing the most basic setup now!

**Comment by tech4GT** [comment]
@jywarren Can you please grant me push access to the repository :sweat_smile:

**Comment by jywarren** [comment]
doing so now, thanks!!!

**Comment by tech4GT** [comment]
Okay I'll try to deploy the container with a very basic setup tomorrow, and then I'll raise an issue for the optimizations, maybe I can document some of my ideas over there too!
Also on a different note I tried out the app locally and it works like a charm :v:

**Comment by jywarren** [comment]
oh wow!!! very cool.

Check out the various "projects" several of which are optimization related:
https://github.com/publiclab/image-sequencer/labels/project

**Comment by tech4GT** [comment]
One think I am concerned about though is, if we do switch to web assembly, what parts of the main code we would need to re-write or should we just switch to something like openCV entirely?
I think we can start with making optimizations in javascript and then move towards web-assembly if that gets unmanageable, what do you think?

**Comment by jywarren** [comment]
Yeah i am not sure about it. I think we can follow multiple paths to
optimize and should probably discuss that in the IS repo. Switching several
modules to openCV would be powerful and flexible. So would webAssembly of
pixelManipulation.

**Comment by tech4GT** [comment]
I think you are right, also please do have a look at the repository, I have pushed the basic file I wrote earlier today, will be extending this **A LOT** but I think this gives us a start.

**Comment by icarito** [comment]
Just a note that Google Cloud Engine has Standard Environment and Flexible Environment and Ruby seems to only be supported on Flexible Environment which is significantly more expensive: https://cloud.google.com/appengine/docs/standard/appengine-generation

**Comment by tech4GT** [comment]
Oh, but is-app will be pure node.js, so I guess thats not a problem!

**Comment by jywarren** [comment]
Hi @tech4GT @icarito @sashadev-sky and others -- i just uploaded diagrams above in which i tried to very clearly articulate the current and planned export workflows. Please have a look!

**Comment by jywarren** [comment]
However, we should think about, in both cases, what points we should try to report status in a `status.json` file which could be polled in JavaScript by MapKnitter users as their export runs, to be able to see what stage their work is in. 

This, and other aspects such as the parallel running and the image pairing during compositing, make me think we really need to consider a new layer, a `mapknitter-exporter-runner` that could persist a bit longer, run in a container itself, but could persist a `status.json` file for the entire export run. 

We could even think more broadly and develop it as an `image-sequencer-runner` which can handle complex branching image sequencer runs. @tech4GT maybe this is where the full express-based `image-sequencer-app` comes in, since the simpler individual steps seem to be possible using just cloud functions? Love to hear your thoughts on all this.

**Comment by jywarren** [comment]
Haha awesome label @sashadev-sky -- i'll respond more completely later today i hope!

Just noting that @icarito has created a Dockerfile for the GDAL/ImageMagick container track: https://github.com/publiclab/mapknitter/pull/349

**Comment by tech4GT** [comment]
Hi @jywarren 
I was thinking about the is-runner and maybe we can base it on the nodejs clustered api?
https://nodejs.org/api/cluster.html
Also How do we want to divide up the work inside these processes exactly? I mean is it specified by the user or we want some kind of algorithm to decide?

**Comment by jywarren** [comment]
Hi Varun!!! Interesting. I guess we could start by noting for each step
what previous steps must be complete for it to run, and we could point each
at prior step references. We might track their state, but also trigger a
re-assessment of if all are complete, using an event listener? I think it
might be worth writing this out step by step. Like:

1. receive multiple image URLs and coordinates
2. step one: run one process for each image, /concurrently/
3. once all images are done with step 2, begin next step
4. go through images one by one and add them /sequentially/ to the previous
image (using coordinate offsets) to make a big combined image
5. ...

at some of these steps, we would need to know a) what triggers the step to
check if it can begin, and b) what conditions must be met for it to start,
right?

**Comment by tech4GT** [comment]
This makes sense Jeff, let me write up some code and see if it works, let’s build this into is-app.

**Comment by jywarren** [comment]
That's great. You could even start with a sequence that doesn't yet do
distortion, but something simpler that already works. Then we'll have the
shell of the system in the right format and can focus on just getting the
internal modules to work.

Thanks!

**Comment by tech4GT** [comment]
Yeah that’s what I was thinking, we can plug in the distortion part later since there are a couple of options I need to explore there and I don't want to slow this down because of that!

**Comment by tech4GT** [comment]
@jywarren Is there any is-module which stitches the images together? Or do I need to write one?

**Comment by tech4GT** [comment]
Also, the information about the steps, as to the complete graph, do we get that through JSON somehow?
I was thinking we can sort of a topological sort on the dependency graph which will automatically run our steps in the right order, what do you think?
@jywarren

**Comment by tech4GT** [comment]
I started this and I have implemented steps 1 and 2. ie we get an array of urls in the request. We apply the preliminary step to all images.(Currently I have done it using async calls but later we can move to clustered API as well if we decide to deploy to a more powerful system)
Now I need the exact information as to in what order what steps need to be applied to these images and A way to stitch them together so we can export our final result. :smile:

**Comment by jywarren** [comment]
For now we could use the overlay module perhaps, although we might need to
make a change or add a new module to change the image dimensions so it
doesn't get cut off, right? How does that sound? They'll then be composited
although not yet distorted.

**Comment by tech4GT** [comment]
Hmm, I think we will need a stitch module for this down the road but I can try to make do with overlay. I'll try it out and let you know.
Also what about the details of the order in which the processing will happen. I think I explained in one of the previous comments, does the json contain the exact order of processing images??

**Comment by jywarren** [comment]
Agreed, we can use overlay for now but plan to do a more efficient stitch
module later. I think there are 2 paths for ordering - for the
compositing/overlay step, we should do it in the order that's given in the
original JSON, yes. (later we could do this in a tree diagram, partially
parallelized like in my diagrams, but we can do the easier linear version
in the first implementation)

For the warping/distorting step, the order shouldn't matter. Thanks!

**Comment by tech4GT** [comment]
@jywarren I have pushed the initial code in the is-app repo, another problem we have to rectify is Import-Image, it uses `new Image()` and hence cannot be used in node and without that we cannot use the overlay module. I guess we would have to go for the knit module afterall :sweat_smile:

**Comment by tech4GT** [comment]
Also I was meaning to ask whether leaflet works in node? Actually I have no idea about that. Does it use gl?

**Comment by jywarren** [comment]
For the coordinate conversions? Hmm I'm not sure, but if it doesn't we can
probably use proj4.js.

We should make sure import-image also works in node, huh? But not as part
of this project unless it's the way to advance our compositing... I'd guess
it is, though... Why isn't import image using the same method as the
initial module? Probably for efficiency, I guess, if new Image is faster.

**Comment by tech4GT** [comment]
Okay so do you want me to update import image to work in node using method similar to lead-image then?

**Comment by jywarren** [comment]
If that seems like the best way forward, what do you think?

**Comment by tech4GT** [comment]
I mean without that there's no way of merging images anyway, so let's go with this for now!

**Comment by tech4GT** [comment]
Okay and another thing is that we need a way to figure out where exactly to overlay the images onto one another, as in if some Image needs to be rotated first and such. Would it be possible for you to give me a simple sample to work this out, actually I tried doing this with the test dataset you mentioned but it's kind of confusing. I was wondering If I can get a stripped down version of that data on which I can figure this out.

**Comment by jywarren** [comment]
Well, rotation will be done as part of the warping, not as a separate step.
So I think best intermediate approach would be like just placing on the
first of the corner points provided. There should be code in the exporter
library to calculate relative positions -- let me see:

https://github.com/publiclab/mapknitter-exporter/blob/main/lib/mapknitterExporter.rb#L252

Actually this newly spun out library does the whole sequence in the
traditional ruby gdal toolchain. It's a good reference for us on the
Javascript version!

**Comment by tech4GT** [comment]
I'll check this out!

**Comment by tech4GT** [comment]
Alright, so we have import Image working now, I'll try a workflow of combining Images, right now overlay cuts the image off but we can fix that in a stitch module later.

**Comment by jywarren** [comment]
Yes or perhaps we can make a "resize-canvas" or "canvas-size" module; that might be useful later anyways.

**Comment by jywarren** [comment]
Portions of the above in the Ruby workflow are coming online:

* https://github.com/publiclab/mapknitter-exporter-sinatra - Ruby app!
* https://github.com/publiclab/mapknitter-exporter - Ruby gem!

Thanks @icarito !!!

**Comment by tech4GT** [comment]
@jywarren Since divy is looking into puppeteer for the gl based implementation, I was wondering what should we pick next?
I mean on the js track we are left with canvas-resize, geotiff and JSON conversion. Did I miss anything?

**Comment by jywarren** [comment]
Those are definitely critical, so let's dive into canvas-resize and
Geotiff? And if you can keep supporting Divy on webgl too, that would be
awesome. Thank you Varun!

**Comment by tech4GT** [comment]
Sure! I’ll get started of cancas-resize then. 😄

**Comment by tech4GT** [comment]
Okay just to sum up, we have a basic exporter process running in gCloud based on is-app
Now the nest steps would be:
- To complete the function which converts the lat/lon coordinates in the json to the inputs for the sequencer module
- Export the completed app as a docker container and deploy it on cloud engine
- Add a geotiff generation module in image sequencer
- Doing optimizations to make the export process faster.
cc @jywarren @icarito

**Comment by jywarren** [comment]
This seems just right, although geotiff can wait, for sure! I think the key here is getting the canvas-resize, the overlay/compositing, and the whole thing running to generate a correct output for a given map, which need not include geotiff for now. And actually the optimization is important, but with more memory, we should be OK on that too, and can circle back to that later. 

For reference, here's where we're testing the output of map exports in the Ruby-based parallel exporter project: https://github.com/publiclab/mapknitter-exporter-sinatra/issues/23

Those maps and the `warpables.json` files they're based on are great for demonstrating that this system is working properly and for comparing the output to what MapKnitter shows!

**Comment by tech4GT** [comment]
Update on the node.js exporter:
- The export process is now using multiSequencer parallelized workflow!
- The app is set up with docker, and is ready to be deployed and tested.
cc @jywarren

**Comment by jywarren** [comment]
OK, awesome. Great to hear it!

**Comment by tech4GT** [comment]
@jywarren I was also thinking of adding a benchmarking test like we have on main is repository. That'll tell us how each of the optimizations we make affects the server response times!

One more thing.. what kind of workload can we expect on this service? I mean is it possible we would be dealing with multiple requests at the same time?
Actually, right now I have not scaled the app for multiple processors, and it will depend on our load how we want to scale.
So if we can have concurrent requests then I recommend using cluster API to mimic the server instance on all the cores.
And if we are looking at single request at a time we can run the different steps on different cores to make the process even faster?
:smile:

**Comment by jywarren** [comment]
And, we still need to address the array size issue in get-pixels and
save-pixels, right? I made an issue here to track all of that thoroughly:
https://github.com/publiclab/image-sequencer-app/issues/12

**Comment by tech4GT** [comment]
@jywarren Could you please also respond to the query I mentioned above! Thanks a ton! :)

**Comment by jywarren** [comment]
We should be able to load balance a cluster using the infrastructure Sebastian has made, yes!

**Comment by tech4GT** [comment]
Oh! In that case I don't have to worry about it! Awesome!! Thanks a lot!! 🤩

**Comment by jywarren** [comment]
:-)

**Comment by jywarren** [comment]
Getting closer in https://github.com/publiclab/mapknitter/pull/1192
